### ISSUE TITLE
Add proper wrappers for vector types and make names consistent and intuitive

### ISSUE THREAD

**Issue by clauswilke** [extendr/extendr#266]
Related to marshaling rules (PR #261) and consistent wrapper types for R scalars (#264): Our handling of R integer, real, or logical vectors is currently confusing and inconsistent. I think we need to clean up names and add a few more wrappers. These seem like two issues but are really just one.

First, we have numerous proper wrapper types (e.g., `List`):
https://github.com/extendr/extendr/blob/00fed0d8fce2ca0a3c4bf822f685d0d6ada0a1ae/extendr-api/src/wrapper/list.rs#L3-L6
but we don't have an equivalent for R vectors of integers, doubles, characters, etc. Those are currently handled as iterators:
https://github.com/extendr/extendr/blob/00fed0d8fce2ca0a3c4bf822f685d0d6ada0a1ae/extendr-api/src/iter.rs#L76-L84
I think this is confusing and difficult to reason about.

Second, confusion is increased by the fact that the naming doesn't help with understanding what we're dealing with. `List` is a wrapper object, but `Real` is an iterator. I have to memorize this and can't tell from the name.

Therefore, I propose two tasks:
1. For types that are iterators, add `Iter` to the name. E.g. `IntIter`, `RealIter`.
2. For wrapper types that represent vectors of some scalar type, make the name plural so it's clear they don't represent a scalar value. E.g., `Ints` (or `Integers`?), `Reals`, `Logicals`, `Characters` (or `Strings`?), `Complexes`. This is similar to the [cpp11](https://cran.r-project.org/web/packages/cpp11/vignettes/converting.html) approach.

@Ilia-Kosenkov @andy-thomason @yutannihilation Could you comment, please?

**Comment by yutannihilation** [comment]
I totally support your proposal. I've been thinking we need to overhaul `SliceIter`-variants (c.f. https://github.com/extendr/extendr/issues/251#issuecomment-894748546).

>     2\. E.g., `Ints` (or `Integers`?), `Reals`, `Logicals`

Do they return bare `i32`, `f64`, or `bool`? Or, `Rint`, `Rfloat`, or `Rbool`? If we prefer `Rt`, the names might be `Rints`, `Rfloats`, ..., accordingly. I prefer the bare ones, though.

**Comment by clauswilke** [comment]
I think the names should be consistent with the other wrappers, e.g. `List`, `Pairlist`, etc.

**Comment by yutannihilation** [comment]
Thanks, that's convincing. I have no opinion against the names then.

**Comment by Ilia-Kosenkov** [comment]
Coming from #261, my suggestion is the following:
1. Primitive wrapper types that are binary compatible with `Rust` types should have small and concise names, like `Rint` and `Rfloat`. Note that the names are derived from `Rust` types, so `Rint` for `i32`, `Rfloat` for `f64`, `Rbool` for `bool`. `Rust` requires type names to be PascalCase, so we can discuss whether it should be `Rint` or `RInt`.
2. ALTREPs. No opinion here, these types, for now, are expected to be internal to `extendr`. If inspiration is drawn from `R` headers, possible names would be `AltInteger`, `AltLogical`, etc.
3. Plain data iterators. Not sure if in the current paradigm we need separate iterators (I still prefer to have an indexer trait as well). For this, I would prefer to have `*Iter` names, perhaps `IntegerIter`, `LogicalIter`, etc
4. Vector types that are equivalent to `Either<,>`. These types are expected to be used by the consumers, so their names should be friendly. I suggest using that of `{cpp11}`, i.e. `Integers` (plural) is approximate;y `Either<IntegerIter, AltInteger>`. The internals of these types are an implementation detail, so users will work with just `Integers`, just `Logicals`. One exception would be `Strings`, because `Characters` is completely misleading (its not characters, its a collection of `&[str]`)
5. Special discriminated unions to accept numeric vectors of one of the supported types. The first obvious choice is `Numerics = Integers | Doubles`, which is self-explanatory. The other one which I would like to see is `ComplexNumerics = Integers | Doubles | Complexes` (or) `Numerics | Complexes`. The last name can be adjusted.

The main reasoning here is that the obscure types that users will likely work with should mimic names of `R` type system, so `Numerics` means exactly what `nuemric` means in `R`: `is.numeric(1) == is.numeric(1L) == TRUE`. Then, vector item types are primitives, so `Integers`  contain `Rint`s,  `Doubles` contain `Rfloat`s, which in most cases will be immediately unwrapped into a regular `Rust` type.

One issue here is that in user `R` code floats are called `double`s, then in `C` API they are `REALSXP` (so reals), then `Rust` works with floats (`f32`/`f64`). Basically, `double`, `real`, `float64` all represent the same entity, so the choice is difficult. Same applies to `logical`, which effectively contains `bool`, `raw`, which contains `u8` (or `byte`), `character`, which contains `String` (or better `&String`). 

I suggest to look towards `{cpp11}` naming of their iterator types and primitive types.
As one of the examples, their `logicals` actually is a container for `r_bool`, which translates to our `Logicals` implementing `Iterator<Item = Rbool>`:
https://github.com/r-lib/cpp11/blob/138bc2752fdd138d954be738130f98d62bc8563e/inst/include/cpp11/logicals.hpp#L48

Once we finalize naming here, I can re-work #261 so we have something to work with.

**Comment by clauswilke** [comment]
@Ilia-Kosenkov Agreed with everything. I have a slight preference for `Rint`, `Rfloat`, `Rbool`, as `RInt`, `RFloat`, `RBool` seem awkward to me. But I can go either way if somebody else feels strongly about it.

**Comment by andy-thomason** [comment]
For a source of truth for vector names (util.c), we have the following:

```
TypeTable[] = {
    { "NULL",		NILSXP	   },  /* real types */
    { "symbol",		SYMSXP	   },
    { "pairlist",	LISTSXP	   },
    { "closure",	CLOSXP	   },
    { "environment",	ENVSXP	   },
    { "promise",	PROMSXP	   },
    { "language",	LANGSXP	   },
    { "special",	SPECIALSXP },
    { "builtin",	BUILTINSXP },
    { "char",		CHARSXP	   },
    { "logical",	LGLSXP	   },
    { "integer",	INTSXP	   },
    { "double",		REALSXP	   }, /*-  "real", for R <= 0.61.x */
    { "complex",	CPLXSXP	   },
    { "character",	STRSXP	   },
    { "...",		DOTSXP	   },
    { "any",		ANYSXP	   },
    { "expression",	EXPRSXP	   },
    { "list",		VECSXP	   },
    { "externalptr",	EXTPTRSXP  },
    { "bytecode",	BCODESXP   },
    { "weakref",	WEAKREFSXP },
    { "raw",		RAWSXP },
    { "S4",		S4SXP },
    /* aliases : */
    { "numeric",	REALSXP	   },
    { "name",		SYMSXP	   },

    { (char *)NULL,	-1	   }
};
```

These are used by `str2type()` in many places in the source, for instance when creating vectors
from scratch:

```
> double
function (length = 0L) 
.Internal(vector("double", length))
<bytecode: 0x5563ebca9870>
<environment: namespace:base>
```

**Comment by andy-thomason** [comment]
We are quite close to this list with Wrapper names and RType (should this be Rtype?)

**Comment by clauswilke** [comment]
@andy-thomason Agreed on most of these names, with the exceptions as noted in the various discussions:
- Vector-valued types should be plural (integers rather than integer) to remove confusion. This is in line with the approach of the cpp11 package
- Let's not use "character(s)" but instead "string(s)", since "character" for a collection of strings is a poor name and highly confusing.

**Comment by clauswilke** [comment]
And yes, Rtype.

**Comment by andy-thomason** [comment]
Sounds reasonable...

**Comment by andy-thomason** [comment]
I'm keen not to alias the std::String type, but Char or Str are options.

**Comment by clauswilke** [comment]
> I'm keen not to alias the std::String type, but Char or Str are options.

Under my proposal we would have `Rstr` for a single string and `Strings` for a vector of strings. Does this work for you, or is it too close to `std::String`?

**Comment by andy-thomason** [comment]
No, that works fine. An Rstr would wrap a single character object.

I'm moving towards making Robj just a wrapper of a SEXP pointer without the enum.
This will enable us to use slices of Robjs for List and Strings types.

**Comment by andy-thomason** [comment]
We also need to tighten up the mutablility of Robj types.

I've been requiring `&mut self` for anything which can change an Robj. such as set_slot, set_attrib.
This should be true of mutable slices, also.

**Comment by clauswilke** [comment]
Great, we're all in agreement then I believe.

**Comment by andy-thomason** [comment]
It would be worth having a "elt()", "region()" and "set_elt()" methods for each wrapper type.

For ALTREP, these are quite efficient.

**Comment by andy-thomason** [comment]
If you like, I can make a sample PR for "Integers" for discussion.

**Comment by clauswilke** [comment]
That would be great, yes.

**Comment by multimeric** [comment]
This seems to be largely implemented via [[#274]](https://github.com/extendr/extendr/pull/274), [[#284]](https://github.com/extendr/extendr/pull/284), [[#301]](https://github.com/extendr/extendr/pull/301), [[#338]](https://github.com/extendr/extendr/pull/338), and [[#350]](https://github.com/extendr/extendr/pull/350). Can this be closed, or are there still missing types?

**Comment by JosiahParry** [comment]
I'm going to close this issue as I believe this is completely implemented in extendr-api as it is today!
